uk.co.compucorp.civicrm.hrcore/CRM/HRCore/CMSData/Variable/VariableServiceInterface.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * Implement this interface when providing a class to interact with system
+ * variables for a CMS.
+ */
+interface CRM_HRCore_CMSData_Variable_VariableServiceInterface {
```

Comment:
> The interfaces are in different levels

I moved the `PathsInterface` into the same level as the classes that implement it.

> On the first one, the CMS term is only mentioned once in the full class name, while on the second one it appears twice in the Factory name.

Agree, I will rename but might be changed anyway if not using the `VariableService` idea

> Another reason is that I'm afraid that not all CMSes have this concept of variable and if that's really the case

Fair point, but do we really want to investigate all other CMS functionality before starting development on something in CiviHR that depends on Drupal? In my opinion looking through the docs for Wordpress and Joomla! would be a total waste of time right now since we have no plans to be compatible with them.

> Besides that, I really think we abuse the term Service sometimes.

For more specific classes it's easier to come up with unique, descriptive class names. For example if you have a class that is only responsible for creating a contact it's easy to call it `ContactCreator` or something like that. However for classes that do CRUD actions on a certain entity I found in the past that a `ContactService` is a clear name. In this case the table in the database being updated / read was `variable`, so I thought `VariableService` would be consistent with my logic. I know variable is a confusing term here, but I didn't want to break with the naming convention I was using.

You've mentioned a few times in PR reviews that you don't like the "service" suffix, but I really don't agree with your previous proposal that we just call it by the entity name. I like using short classnames because it makes keeping to the line limit a lot easier, and code is more readable. If we omit the service in some cases, as you did with `CRM_HRCore_Service_Manager` and we continue using aliases like `use CRM_HRCore_Service_Manager as Manager` we could have code like `$x = new Manager();`. To anyone reading this it would appear you are creating a new manager entity. In fact I've seen you've aliased this service as `ManagerService` in a [test](https://github.com/civicrm/civihr/commit/84ec3dc05d8a74593b9b796e97337b2ce243768f#diff-cad02b808a5fd40199869ac98aba3a2aR5) which I think suggests the original name wasn't clear enough.

I know `Service` isn't great, if you look at the other classes I created for stats gathering we have `CiviHRStatisticsJSONConvertor` and `CiviHRStatsGatherer`, each with only one public method. I prefer these kinds of classes because they have a single purpose, and because of that are easy to find a name for.

I'm happy to go with the approach of having only a single public method per class, in which case the `VariableService` would have separate classes for getting and setting. I think this is probably more in keeping with the SRP. At the same time I think it's sometimes cleaner to group simple actions in a single class, but we need a name for these kinds of classes.

> What are the difference between these classes and the ones in the CRM_HRCore_Service "namespace"?

There is no difference. I would have put these classes in the Service namespace if it wasn't for the fact that Alessandro had already put CMS related stuff here. The reason I would have put them under "Service" is that I don't want to pollute the top-level namespace CMS namespace too much. Otherwise when we open the CMS directory we could have lots and lots of top-level directories which makes navigation more complicated. If we are moving towards merging other extensions into hrcore then I think we should consider what directory structure we would like to use for it. Maybe we could have a discussion about this with the other backend developers and decide on a structure.

> Finally, I think that the VariableService not very good.

I agree, it's bullshit. It's making an interface for implementations we're never going to use (and don't even know will exist in other CMSs as you mentioned). This made it difficult to make a plan for this. Usually when creating an interface I can look at how X different concrete classes share the same functionality. In this case the spec was to just use some sort of an abstraction to call Drupal functions.

Reply:
> Fair point, but do we really want to investigate all other CMS functionality before starting development on something in CiviHR that depends on Drupal? In my opinion looking through the docs for Wordpress and Joomla! would be a total waste of time right now since we have no plans to be compatible with them.

No we don't want to do that. The problem here is that we are focusing too much on Drupal and probably not abstracting this the right way. For the purpose of this ticket, we don't want to get drupal variables. We want to get information about the site that Drupal happens to store as a variable in the database. That's why I suggested the `SiteInformation` interface, which sounds generic enough to be supported by all CMSes, instead of Variables that are probably Drupal specific.

> If we omit the service in some cases, as you did with CRM_HRCore_Service_Manager and we continue using aliases like use CRM_HRCore_Service_Manager as Manager we could have code like $x = new Manager();. To anyone reading this it would appear you are creating a new manager entity. In fact I've seen you've aliased this service as ManagerService in a test which I think suggests the original name wasn't clear enough.

The alias doesn't have to be exactly the the last part of the full class name. It is just a useful short way to refer to the class. So, for `CRM_HRCore_Service_Manager` we pick only the important bits of it (Service and Manager) and alias it to `ManagerService`. If we were using PSR4, I would probably do something like this:

```php
use CRM\HRCore\Service as Service;
$managerService = new Service\Manager();
```

Anyway, as I've mentioned in the previous PR, I really don't have any strong opinions about this. It's fine to use the suffix. The problem here is that we have services that are outside CRM_HRCore_Services. This is confusing.

Reply:
> That's why I suggested the SiteInformation interface, which sounds generic enough to be supported by all CMSes, instead of Variables that are probably Drupal specific.

Agree, my implementation was a bit of a hack at the end purely with the purpose of abstracting a Drupal function call, so I didn't really think it through. I'll implement the `SiteInformation` suggestion.

Your example of PSR-4 is fine, although it's not what I'm used to. I see the specific classes being imported more frequently than a whole namespace level, which would be

```
use CRM\HRCore\Service\ManagerService;
$managerService = new ManagerService();
```

This has the advantage that you don't need to use aliases. This could be another topic for group discussion, but not until we talk about using PSR-4 (or even PSR-0 namespaces) first.